### PR TITLE
Prevent crashing dyn files from reopening in automatic run mode

### PR DESCRIPTION
**Pull request by pboyer**
### Issue

We need to prevent files that cause a crash from being reopened with run auto.  This would cause Dynamo to immediately crash on opening the file, effectively preventing the user from ever accessing the file again.

### Fix

Introduce a simple flag, `HasRunWithoutCrash`.  There are several state transitions for this flag, encoded in a new unit test file `CrashProtectionTests`. 

* For files what have never been opened, `HasRunWithoutCrash = false`
	- *So, all files from previous version of Dynamo will open in manual!*
* On run start, `HasRunWithoutCrash = false` is immediately set
	- This makes sure that if the user has modifies the file during a run that causes a crash, there 
	  file is not erroneously marked as having run without crash. 
* On run complete, `HasRunWithoutCrash = true`
* On open file, if run auto & `HasRunWithoutCrash = false`, set to run manual
* On save, if run auto & `HasRunWithoutCrash = true`, ensure that is saved to the file

### Reviewer

- [ ] @lukechurch

### CODE REVIEW

File: src/DynamoCore/Models/HomeWorkspaceModel.cs [hunk 1]
```diff
@@ -87,19 +108,8 @@ public override void Dispose()
...
-        /// the EngineController (or even LiveRunner), but the engine gets reset in 
-        /// a rather nondeterministic way (for example, when Revit idle thread 
-        /// decides it is time to execute a pre-scheduled engine reset). And it gets 
-        /// done more than once during file open. So that's out. The second best 
-        /// place to store this information is then the WorkspaceModel, where file 
-        /// loading is SUPPOSED TO BE done. As of now we let DynamoModel sets the 
-        /// loaded data (since it deals with loading DYN file), but in near future,
-        /// the file loading mechanism will be completely moved into WorkspaceModel,
-        /// that's the time we removed this property setter below.
+        /// In near future, the file loading mechanism will be completely moved
```
Comment (pboyer) [inline src/DynamoCore/Models/HomeWorkspaceModel.cs]: Feels a bit like a tirade than an informative comment.

File: test/DynamoCoreTests/DynamoModelTestBase.cs [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+using Dynamo.ViewModels;
+using DynamoShapeManager;
+using NUnit.Framework;
+using ProtoCore.Mirror;
+using TestServices;
+
+namespace Dynamo
+{
+    public class DynamoModelTestBase : UnitTestBase
+    {
```
Comment (pboyer) [inline test/DynamoCoreTests/DynamoModelTestBase.cs]: We really needed this.  Many of our tests unnecessarily also create a `DynamoViewModel`.

File: test/DynamoCoreTests/CrashProtectionTests.cs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using Dynamo.Models;
+using NUnit.Framework;
+
+namespace Dynamo.Tests
+{
+    [TestFixture]
```
Comment (pboyer) [inline test/DynamoCoreTests/CrashProtectionTests.cs]: This file encodes all of the state transitions indicated above.  All of the tests succeed.

image

**Comment by lukechurch** [comment]
It's a little concerning that this depends on the file not ever getting into the state of Run Auto == true + HasRunOk == true. If it does, it won't recover, as it doesn't automatically persist the flag.

However, without a full 'recovery mode' file implementation I don't see a way round that.

**Comment by lukechurch** [comment]
lgtm

File: test/DynamoCoreTests/DynamoModelTestBase.cs (hunk 2 above)
Comment (ikeough) [inline test/DynamoCoreTests/DynamoModelTestBase.cs]: +1

**Pull request merged**